tracer/src/Datadog.Trace/AppSec/Waf/Waf.cs

```diff
@@ -51,33 +53,87 @@ internal Waf(IntPtr wafHandle, WafLibraryInvoker wafLibraryInvoker)
         /// <param name="embeddedRulesetPath">can be null, means use rules embedded in the manifest </param>
         /// <param name="rulesFromRcm">can be null. RemoteConfig rules json. Takes precedence over rulesFile </param>
         /// <param name="setupWafSchemaExtraction">should we read the config file for schema extraction</param>
+        /// <param name="useUnsafeEncoder">use legacy encoder</param>
         /// <returns>the waf wrapper around waf native</returns>
-        internal static InitResult Create(WafLibraryInvoker wafLibraryInvoker, string obfuscationParameterKeyRegex, string obfuscationParameterValueRegex, string? embeddedRulesetPath = null, JToken? rulesFromRcm = null, bool setupWafSchemaExtraction = false)
+        internal static unsafe InitResult Create(WafLibraryInvoker wafLibraryInvoker, string obfuscationParameterKeyRegex, string obfuscationParameterValueRegex, string? embeddedRulesetPath = null, JToken? rulesFromRcm = null, bool setupWafSchemaExtraction = false, bool useUnsafeEncoder = false)
         {
             var wafConfigurator = new WafConfigurator(wafLibraryInvoker);
 
             // set the log level and setup the logger
             wafLibraryInvoker.SetupLogging(GlobalSettings.Instance.DebugEnabledInternal);
+
             var jtokenRoot = rulesFromRcm ?? WafConfigurator.DeserializeEmbeddedOrStaticRules(embeddedRulesetPath)!;
 
-            var argCache = new List<Obj>();
-            var configObj = Encoder.Encode(jtokenRoot, wafLibraryInvoker, argCache, applySafetyLimits: false);
+            IntPtr rulesObj;
+            List<Obj>? legacyArgsToDispose = null;
```

Comment:
lol yes it's because legacy was disposing all the tree of arguments whereas the new encoder is recycling internally, but with the abstractions on top now it's more seamless: 28d8340d138e09f7b3d097899ac9a0a4ad7e26fa